### PR TITLE
SoundWire integration

### PR DESCRIPTION
**Pull request by plbossart**
This PR gathers all the known SoundWire patches on a single branch.
This is not ready to be merged just yet but we do want CI results and QA tests on this branch

Most of the patches in this branch have been reviewed multiple times, but additional pair of eyes would not hurt (some of the patches will be submitted directly upstream)

All other existing branches related to SoundWire will be archived.

### CODE REVIEW

**Comment by plbossart** [comment]
@xiulipan @fredoh9 nothing tested except Cyan?

**Comment by xiulipan** [comment]
@plbossart I see all platforms are failed due to
```
Removing soundwire_bus
rmmod: ERROR: Module soundwire_bus is in use by: soundwire_generic_allocation
```

If that means we need to update the kmod scripts used by CI? [kmod scripts](https://github.com/thesofproject/sof-test/tree/master/tools/kmod) 

PS: I create pr to clean up test scripts in SOF with https://github.com/thesofproject/sof/pull/2274

**Comment by plbossart** [comment]
> @plbossart I see all platforms are failed due to
> 
> ```
> Removing soundwire_bus
> rmmod: ERROR: Module soundwire_bus is in use by: soundwire_generic_allocation
> ```
> 
> If that means we need to update the kmod scripts used by CI? [kmod scripts](https://github.com/thesofproject/sof-test/tree/master/tools/kmod)
> 
> PS: I create pr to clean up test scripts in SOF with [thesofproject/sof#2274](https://github.com/thesofproject/sof/pull/2274)

well yes, this module is handled in sof/tools/kmod_scripts, so if you don't track those changes it's not going to work.. I think it's high-time we stop the duplication and have a single tool that's used.

**Comment by xiulipan** [comment]
@plbossart new patch added to sof-test.
Will try to re-run the test after the pr is merged.

File: drivers/soundwire/bus.c [hunk 1]
```diff
@@ -456,36 +547,51 @@ static int sdw_get_device_num(struct sdw_slave *slave)
 static int sdw_assign_device_num(struct sdw_slave *slave)
 {
 	int ret, dev_num;
+	bool new_device = false;
+
+	dev_dbg(slave->bus->dev, "in %s\n", __func__);
+
+
+	dev_dbg(slave->bus->dev, "in %s\n", __func__);
```
Comment (bardliao) [inline drivers/soundwire/bus.c]: duplicate dev_dbg

File: drivers/soundwire/generic_bandwidth_allocation.c [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+{
+	struct sdw_slave_runtime *s_rt = NULL;
+	struct sdw_port_runtime *p_rt;
+	int port_bo, sample_int;
+	unsigned int rate, bps, ch = 0;
+
+	port_bo = t_data->block_offset;
+
+	list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
+		rate = m_rt->stream->params.rate;
```
Comment (RanderWang) [inline drivers/soundwire/generic_bandwidth_allocation.c:49]: @plbossart   As we discussed,  "port_bo = t_data->block_offset;" should be moved into loop to support two slaves on single link in one stream. In this case, all slaves will work on the same dp offset. And we can set mixer to each slave to change its pattern

I tested on TGL with two 1308s on link1.

File: drivers/soundwire/cadence_master.c [hunk 3]
```diff
@@ -1178,6 +1249,189 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+bool sdw_cdns_is_clock_stop(struct sdw_cdns *cdns)
+{
+	u32 status;
+
+	status = cdns_readl(cdns, CDNS_MCP_STAT) & CDNS_MCP_STAT_CLK_STOP;
+	if (status) {
+		dev_dbg(cdns->dev, "Clock is stopped\n");
+		return true;
+	}
+
```
Comment (bardliao) [inline drivers/soundwire/cadence_master.c:1264]: Would it be better to add a dev_dbg here to say Clock is not stopped?

**Comment by RanderWang** [comment]
@Pierre,  Today, I only reviewed the patches I involved. It is good for me.

File: drivers/soundwire/intel.c [hunk 4]
```diff
@@ -88,12 +105,10 @@
 #define SDW_ALH_STRMZCFG(x)		(0x000 + (0x4 * (x)))
 #define SDW_ALH_NUM_STREAMS		64
 
-#define SDW_ALH_STRMZCFG_DMAT_VAL	0x3
+#define SDW_ALH_STRMZCFG_DMAT_VAL	0xf
```
Comment (bardliao) [inline drivers/soundwire/intel.c]: As https://github.com/thesofproject/sof/pull/2226 is merged, I guess we don't need this change now

File: drivers/soundwire/intel.c [hunk 5]
```diff
@@ -880,6 +1274,7 @@ static int sdw_master_read_intel_prop(struct sdw_bus *bus)
 				 "intel-sdw-ip-clock",
 				 &prop->mclk_freq);
 
+	prop->mclk_freq /= 2;
```
Comment (bardliao) [inline drivers/soundwire/intel.c]: Is it better to have commit to describe why ``/= 2`` is required?

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: removed both

File: drivers/soundwire/intel.c (hunk 5 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: done

File: drivers/soundwire/intel.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: will remove, thanks

**Comment by xiulipan** [comment]
SOFCI TEST

**Comment by xiulipan** [comment]
@plbossart @mengdonglin run a new round test.
Expect for the known random error with dsp reset on ICL
```
Caught error: Jan 13 23:02:40 jf-icl-rvp-hda-4 kernel: [  409.811379] sof-audio-pci 0000:00:1f.3: error: cl_dsp_init: timeout HDA_DSP_SRAM_REG_ROM_STATUS read
Jan 13 23:02:40 jf-icl-rvp-hda-4 kernel: [  409.811426] sof-audio-pci 0000:00:1f.3: error: status = 0x00000000 panic = 0x00000000
Jan 13 23:02:40 jf-icl-rvp-hda-4 kernel: [  409.811472] sof-audio-pci 0000:00:1f.3: iteration 0 of Core En/ROM load failed: -110
```

I see new issue with runtime PM. The platform is not idle after 5s delay, but it seems can enter runtime suspend after around 6s.
https://sof-ci.01.org/linuxpr/PR1692/build2886/devicetest/CML_MANTIS_HDA/check-runtime-pm-status.sh/dmesg.txt

@plbossart Should we modify the wait time to fit this case or the longer time to enter suspend is not accepted?
```
kernel: [  419.429558]  HDA Analog: ASoC: BE playback disconnect check for Analog Playback and Capture
kernel: [  419.429561]  HDA Analog: freed DSP playback path HDA Analog -> Analog Playback and Capture
kernel: [  420.043684] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
kernel: [  420.043725] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
kernel: [  424.551584]  HDA Analog: ASoC: pop wq checking: Playback status: inactive waiting: yes
kernel: [  424.766539] sof-audio-pci 0000:00:1f.3: snd_sof_runtime_idle
kernel: [  424.766550] sof-audio-pci 0000:00:1f.3: snd_sof_runtime_suspend
kernel: [  424.766554] sof-audio-pci 0000:00:1f.3: sof_suspend
kernel: [  424.766569] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x20140000 successful
kernel: [  424.801549] sof-audio-pci 0000:00:1f.3: sof_send_pm_ctx_ipc
kernel: [  424.801558] sof-audio-pci 0000:00:1f.3: ipc tx: 0x40010000: GLB_PM_MSG: CTX_SAVE
kernel: [  424.801737] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x40010000: GLB_PM_MSG: CTX_SAVE
kernel: [  424.801841] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x1010f0f successful
kernel: [  424.801846] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0xf0f successful
kernel: [  424.801849] sof-audio-pci 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
kernel: [  424.801884] sof-audio-pci 0000:00:1f.3: Debug PCIR: 00000010 at  00000044
```

**Comment by xiulipan** [comment]
@plbossart It seem the runtime PM issue is from the sof-dev code base.
@ClarexZhou Could you help to confirm the if the PM issue is caused by the time delay increased?

**Comment by ClarexZhou** [comment]
> @plbossart It seem the runtime PM issue is from the sof-dev code base.
> @ClarexZhou Could you help to confirm the if the PM issue is caused by the time delay increased?

Runtime PM status will changed to suspend after about 5.5s. And yes, this fail is caused by the time delay increased

**Comment by xiulipan** [comment]
@plbossart @ClarexZhou hot fix for test case is now send https://github.com/thesofproject/sof-test/pull/42

@ranj063 @bardliao @keyonjie 
I will like to ask about what would be expected behavior for the runtime pm now? I think we get a delay setting to 2s?

**Comment by plbossart** [comment]
> @plbossart @ClarexZhou hot fix for test case is now send [thesofproject/sof-test#42](https://github.com/thesofproject/sof-test/pull/42)
> 
> @ranj063 @bardliao @keyonjie
> I will like to ask about what would be expected behavior for the runtime pm now? I think we get a delay setting to 2s?

Sorry I am unable to figure out what you are asking, and why PM issues on an HDaudio platforms would be introduced by this PR?

**Comment by ranj063** [comment]
> > @plbossart @ClarexZhou hot fix for test case is now send [thesofproject/sof-test#42](https://github.com/thesofproject/sof-test/pull/42)
> > @ranj063 @bardliao @keyonjie
> > I will like to ask about what would be expected behavior for the runtime pm now? I think we get a delay setting to 2s?
> 
> Sorry I am unable to figure out what you are asking, and why PM issues on an HDaudio platforms would be introduced by this PR?

yes, me too. I am not following the question.

**Comment by ClarexZhou** [comment]
Manually test sof-dev (commit: 075b242) + PR#1692 on SDW TGL, CML-U and CML-H products and TGL HDA. Find only one issue on TGL HDA, see #1711 . 

CI test on other platforms also done. No side effect.  Find CI test result in https://sof-ci.01.org/linuxpr/PR1692/build2895/devicetest/

**Comment by keqiaozhang** [comment]
Legal scan is clean for this PR.

**Comment by plbossart** [comment]
This PR has been tested multiple times, most patches are upstream so let's go ahead a merge before the long week-end (MLK Day) in the US and Chinese NewYear, that leaves us several days to fix things if needed

**Pull request merged**